File: backend/neural/layer.py
"""
Layer Implementation for CAMBRIC LABS

A layer contains multiple neurons that process inputs in parallel.
Each layer has an input dimension, output dimension (number of neurons),
and an activation function.

Architecture:
    Layer.forward() -> calls Neuron.forward() for each neuron
    Layer.backward() -> calls Neuron.backward() for each neuron (computes gradients only)
    Layer.update() -> calls Neuron.update() for each neuron (applies gradients)
"""

import numpy as np
from typing import List, Dict, Any, Optional
from .neuron import Neuron, GradientResult


class Layer:
    """
    A dense (fully-connected) layer of neurons.
    
    All neurons in a layer receive the same inputs. Each neuron has its
    own weights and bias. The layer outputs a vector of values, one per neuron.
    
    Design:
    - backward() computes gradients but does NOT update weights
    - update() applies gradients to all neurons
    """
    
    def __init__(
        self,
        name: str,
        input_dim: int,
        output_dim: int,
        activation: str = 'relu',
        seed: Optional[int] = None
    ):
        """
        Initialize a layer with multiple neurons.
        
        Args:
            name: Layer identifier
            input_dim: Number of inputs per neuron
            output_dim: Number of neurons in this layer
            activation: Activation function name
            seed: Random seed for weight initialization
        """
        self.name = name
        self.input_dim = input_dim
        self.output_dim = output_dim
        self.activation = activation
        
        # Create neurons
        self.neurons: List[Neuron] = []
        for i in range(output_dim):
            # Use seed + i for reproducible but different initializations
            neuron_seed = seed + i if seed is not None else None
            self.neurons.append(Neuron(
                input_count=input_dim,
                activation=activation,
                seed=neuron_seed
            ))
        
        # Internal state for caching
        self._last_inputs: Optional[np.ndarray] = None
        self._last_outputs: Optional[np.ndarray] = None
        self._last_gradients: Optional[List[GradientResult]] = None
    
    @property
    def parameter_count(self) -> int:
        """Total trainable parameters in this layer."""
        return sum(n.parameter_count for n in self.neurons)
    
    @property
    def weights_matrix(self) -> np.ndarray:
        """All weights as a matrix (output_dim x input_dim)."""
        return np.array([n.weights for n in self.neurons])
    
    @property
    def biases_vector(self) -> np.ndarray:
        """All biases as a vector."""
        return np.array([n.bias for n in self.neurons])
    
    def forward(self, inputs: List[float]) -> Dict[str, Any]:
        """
        Compute forward pass through all neurons in this layer.
        
        Args:
            inputs: List of input values
            
        Returns:
            Dictionary containing:
            - outputs: List of neuron outputs
            - neuron_details: Details for each neuron computation
        """
        if len(inputs) != self.input_dim:
            raise ValueError(
                f"Input dimension ({len(inputs)}) doesn't match "
                f"layer's input_dim ({self.input_dim})"
            )
        
        inputs_arr = np.array(inputs, dtype=np.float64)
        self._last_inputs = inputs_arr.copy()
        
        outputs = []
        neuron_details = []
        
        for i, neuron in enumerate(self.neurons):
            result = neuron.forward(inputs)
            outputs.append(result['output'])
            neuron_details.append({
                'neuron_index': i,
                'weighted_sum': result['weighted_sum'],
                'output': result['output'],
                'contributions': result['contributions'],
                'weights': result['weights'],
                'bias': result['bias']
            })
        
        self._last_outputs = np.array(outputs)
        
        return {
            'outputs': outputs,
            'neuron_details': neuron_details,
            'weight_matrix': self.weights_matrix.tolist(),
            'bias_vector': self.biases_vector.tolist(),
            'activation': self.activation,
            'input_dim': self.input_dim,
            'output_dim': self.output_dim
        }
    
    def backward(self, output_gradients: List[float]) -> Dict[str, Any]:
        """
        Compute gradients for all neurons.
        
        IMPORTANT: This only computes gradients, does NOT update weights.
        Call update() to apply the gradients.
        
        Args:
            output_gradients: Gradients for each neuron output (dL/dy)
            
        Returns:
            Dictionary with input gradients and gradient information
        """
        if self._last_inputs is None:
            raise RuntimeError("Must call forward() before backward()")
        
        if len(output_gradients) != self.output_dim:
            raise ValueError(
                f"Gradient count ({len(output_gradients)}) doesn't match "
                f"output_dim ({self.output_dim})"
            )
        
        input_gradients = np.zeros(self.input_dim)
        neuron_gradients = []
        
        for i, (neuron, gradient) in enumerate(zip(self.neurons, output_gradients)):
            grad_result = neuron.backward(gradient)
            
            # Accumulate input gradients for previous layer
            input_gradients += grad_result.input_gradients
            
            neuron_gradients.append({
                'neuron_index': i,
                'weight_gradients': grad_result.weight_gradients.tolist(),
                'bias_gradient': grad_result.bias_gradient,
                'input_gradients': grad_result.input_gradients.tolist(),
                'activation_gradient': grad_result.activation_gradient
            })
        
        self._last_gradients = [n.backward(g) for n, g in 
                                zip(self.neurons, output_gradients)]
        
        return {
            'input_gradients': input_gradients.tolist(),
            'neuron_gradients': neuron_gradients,
            'parameter_count': self.parameter_count
        }
    
    def compute_gradients(self, output_gradients: List[float]) -> List[GradientResult]:
        """
        Compute and return gradients for inspection.
        
        Does NOT modify any parameters.
        
        Args:
            output_gradients: Gradients for each neuron output
            
        Returns:
            List of GradientResult objects (one per neuron)
        """
        if self._last_inputs is None:
            raise RuntimeError("Must call forward() before compute_gradients()")
        
        gradients = []
        for neuron, gradient in zip(self.neurons, output_gradients):
            gradients.append(neuron.backward(gradient))
        
        return gradients
    
    def update(
        self, 
        learning_rate: float,
        clip_gradients: Optional[float] = None
    ) -> Dict[str, Any]:
        """
        Apply accumulated gradients to update parameters.
        
        Args:
            learning_rate: Learning rate for gradient descent
            clip_gradients: Optional gradient norm clip
            
        Returns:
            Dictionary with update information
        """
        if self._last_gradients is None:
            raise RuntimeError("Must call backward() before update()")
        
        weight_updates = []
        bias_updates = []
        
        for i, (neuron, grad) in enumerate(zip(self.neurons, self._last_gradients)):
            result = neuron.update(grad, learning_rate, clip_gradients)
            
            weight_updates.append({
                'neuron_index': i,
                'old_weights': result['old_weights'],
                'new_weights': result['new_weights'],
                'changes': result['weight_changes'],
                'gradient_norm': result['gradient_norms']['weight_norm']
            })
            
            bias_updates.append({
                'neuron_index': i,
                'old_bias': result['old_bias'],
                'new_bias': result['new_bias'],
                'change': result['bias_change'],
                'gradient': result['gradient_norms']['bias_gradient']
            })
        
        return {
            'weight_updates': weight_updates,
            'bias_updates': bias_updates,
            'parameters_updated': self.parameter_count,
            'learning_rate': learning_rate
        }
    
    def get_state(self) -> Dict[str, Any]:
        """Get current state of the layer."""
        return {
            'name': self.name,
            'input_dim': self.input_dim,
            'output_dim': self.output_dim,
            'activation': self.activation,
            'parameter_count': self.parameter_count,
            'neurons': [n.get_state() for n in self.neurons],
            'weights_matrix': self.weights_matrix.tolist(),
            'biases_vector': self.biases_vector.tolist()
        }
    
    def set_neuron_weights(self, neuron_index: int, weights: List[float]):
        """Set weights for a specific neuron."""
        self.neurons[neuron_index].set_weights(weights)
    
    def set_neuron_bias(self, neuron_index: int, bias: float):
        """Set bias for a specific neuron."""
        self.neurons[neuron_index].set_bias(bias)
    
    def set_weights_matrix(self, weights: List[List[float]]):
        """Set all weights as a matrix."""
        for i, row in enumerate(weights):
            self.neurons[i].set_weights(row)
    
    def set_biases_vector(self, biases: List[float]):
        """Set all biases as a vector."""
        for i, bias in enumerate(biases):
            self.neurons[i].set_bias(bias)
    
    def reset_cache(self):
        """Reset temporary cache (preserves learned parameters)."""
        self._last_inputs = None
        self._last_outputs = None
        self._last_gradients = None
        for neuron in self.neurons:
            neuron.reset_cache()
    
    def reset_parameters(self, seed: Optional[int] = None):
        """Reset all parameters to new random values."""
        for i, neuron in enumerate(self.neurons):
            neuron_seed = seed + i if seed is not None else None
            neuron.reset_parameters(neuron_seed)
    
    def reset(self):
        """Reset internal state (alias for reset_cache)."""
        self.reset_cache()
    
    def __repr__(self) -> str:
        return (
            f"Layer('{self.name}', input={self.input_dim}, "
            f"output={self.output_dim}, activation={self.activation})"
        )

File: frontend/src/engine/layer.ts
/**
 * Layer Implementation for CAMBRIC LABS
 *
 * Ported 1:1 from backend/neural/layer.py.
 *
 * A layer contains multiple neurons that process the same inputs in
 * parallel. Each neuron has its own weights and bias.
 *
 * Layer.forward()  -> Neuron.forward() for each neuron
 * Layer.backward() -> Neuron.backward() for each neuron (gradients only)
 * Layer.update()   -> Neuron.update() for each neuron (applies gradients)
 */

import { ActivationName, GradientResult, Neuron } from './neuron';

export interface LayerForwardResult {
  outputs: number[];
  neuronDetails: Array<{
    neuronIndex: number;
    weightedSum: number;
    output: number;
    contributions: number[];
    weights: number[];
    bias: number;
  }>;
  weightMatrix: number[][];
  biasVector: number[];
  activation: ActivationName;
  inputDim: number;
  outputDim: number;
}

export interface LayerBackwardResult {
  inputGradients: number[];
  neuronGradients: Array<{
    neuronIndex: number;
    weightGradients: number[];
    biasGradient: number;
    inputGradients: number[];
    activationGradient: number;
  }>;
  parameterCount: number;
}

export interface LayerUpdateResult {
  weightUpdates: Array<{
    neuronIndex: number;
    oldWeights: number[];
    newWeights: number[];
    changes: number[];
    gradientNorm: number;
  }>;
  biasUpdates: Array<{
    neuronIndex: number;
    oldBias: number;
    newBias: number;
    change: number;
    gradient: number;
  }>;
  parametersUpdated: number;
  learningRate: number;
}

export interface LayerState {
  name: string;
  inputDim: number;
  outputDim: number;
  activation: ActivationName;
  parameterCount: number;
  neurons: ReturnType<Neuron['getState']>[];
  weightsMatrix: number[][];
  biasesVector: number[];
}

export class Layer {
  readonly name: string;
  readonly inputDim: number;
  readonly outputDim: number;
  activation: ActivationName;
  neurons: Neuron[];

  private lastInputs: number[] | null = null;
  private lastOutputs: number[] | null = null;
  private lastGradients: GradientResult[] | null = null;

  constructor(opts: {
    name: string;
    inputDim: number;
    outputDim: number;
    activation?: ActivationName;
    seed?: number;
  }) {
    const { name, inputDim, outputDim, activation = 'relu', seed } = opts;

    this.name = name;
    this.inputDim = inputDim;
    this.outputDim = outputDim;
    this.activation = activation;

    this.neurons = Array.from(
      { length: outputDim },
      (_, i) =>
        new Neuron({
          inputCount: inputDim,
          activation,
          seed: seed !== undefined ? seed + i : undefined,
        })
    );
  }

  get parameterCount(): number {
    return this.neurons.reduce((sum, n) => sum + n.parameterCount, 0);
  }

  get weightsMatrix(): number[][] {
    return this.neurons.map((n) => [...n.weights]);
  }

  get biasesVector(): number[] {
    return this.neurons.map((n) => n.bias);
  }

  forward(inputs: number[]): LayerForwardResult {
    if (inputs.length !== this.inputDim) {
      throw new Error(
        `Input dimension (${inputs.length}) doesn't match layer's inputDim (${this.inputDim})`
      );
    }

    this.lastInputs = [...inputs];

    const outputs: number[] = [];
    const neuronDetails: LayerForwardResult['neuronDetails'] = [];

    this.neurons.forEach((neuron, i) => {
      const result = neuron.forward(inputs);
      outputs.push(result.output);
      neuronDetails.push({
        neuronIndex: i,
        weightedSum: result.weightedSum,
        output: result.output,
        contributions: result.contributions,
        weights: result.weights,
        bias: result.bias,
      });
    });

    this.lastOutputs = outputs;

    return {
      outputs,
      neuronDetails,
      weightMatrix: this.weightsMatrix,
      biasVector: this.biasesVector,
      activation: this.activation,
      inputDim: this.inputDim,
      outputDim: this.outputDim,
    };
  }

  /** Computes gradients for all neurons. Does NOT update weights — call update() for that. */
  backward(outputGradients: number[]): LayerBackwardResult {
    if (this.lastInputs === null) {
      throw new Error('Must call forward() before backward()');
    }
    if (outputGradients.length !== this.outputDim) {
      throw new Error(
        `Gradient count (${outputGradients.length}) doesn't match outputDim (${this.outputDim})`
      );
    }

    const inputGradients = new Array(this.inputDim).fill(0);
    const neuronGradients: LayerBackwardResult['neuronGradients'] = [];
    this.lastGradients = [];

    this.neurons.forEach((neuron, i) => {
      const gradResult = neuron.backward(outputGradients[i]);
      this.lastGradients!.push(gradResult);

      for (let j = 0; j < this.inputDim; j++) {
        inputGradients[j] += gradResult.inputGradients[j];
      }

      neuronGradients.push({
        neuronIndex: i,
        weightGradients: gradResult.weightGradients,
        biasGradient: gradResult.biasGradient,
        inputGradients: gradResult.inputGradients,
        activationGradient: gradResult.activationGradient,
      });
    });

    return { inputGradients, neuronGradients, parameterCount: this.parameterCount };
  }

  /** Computes and returns gradients for inspection without modifying parameters. */
  computeGradients(outputGradients: number[]): GradientResult[] {
    if (this.lastInputs === null) {
      throw new Error('Must call forward() before computeGradients()');
    }
    return this.neurons.map((neuron, i) => neuron.backward(outputGradients[i]));
  }

  /** Applies gradients computed by the last backward() call. */
  update(learningRate: number, clipGradients?: number): LayerUpdateResult {
    if (this.lastGradients === null) {
      throw new Error('Must call backward() before update()');
    }

    const weightUpdates: LayerUpdateResult['weightUpdates'] = [];
    const biasUpdates: LayerUpdateResult['biasUpdates'] = [];

    this.neurons.forEach((neuron, i) => {
      const result = neuron.update(this.lastGradients![i], learningRate, clipGradients);

      weightUpdates.push({
        neuronIndex: i,
        oldWeights: result.oldWeights,
        newWeights: result.newWeights,
        changes: result.weightChanges,
        gradientNorm: result.gradientNorms.weightNorm,
      });

      biasUpdates.push({
        neuronIndex: i,
        oldBias: result.oldBias,
        newBias: result.newBias,
        change: result.biasChange,
        gradient: result.gradientNorms.biasGradient,
      });
    });

    return {
      weightUpdates,
      biasUpdates,
      parametersUpdated: this.parameterCount,
      learningRate,
    };
  }

  getState(): LayerState {
    return {
      name: this.name,
      inputDim: this.inputDim,
      outputDim: this.outputDim,
      activation: this.activation,
      parameterCount: this.parameterCount,
      neurons: this.neurons.map((n) => n.getState()),
      weightsMatrix: this.weightsMatrix,
      biasesVector: this.biasesVector,
    };
  }

  setNeuronWeights(neuronIndex: number, weights: number[]): void {
    this.neurons[neuronIndex].setWeights(weights);
  }

  setNeuronBias(neuronIndex: number, bias: number): void {
    this.neurons[neuronIndex].setBias(bias);
  }

  setWeightsMatrix(weights: number[][]): void {
    weights.forEach((row, i) => this.neurons[i].setWeights(row));
  }

  setBiasesVector(biases: number[]): void {
    biases.forEach((bias, i) => this.neurons[i].setBias(bias));
  }

  /** Resets temporary cache; preserves learned parameters. */
  resetCache(): void {
    this.lastInputs = null;
    this.lastOutputs = null;
    this.lastGradients = null;
    this.neurons.forEach((n) => n.resetCache());
  }

  resetParameters(seed?: number): void {
    this.neurons.forEach((n, i) => n.resetParameters(seed !== undefined ? seed + i : undefined));
  }

  reset(): void {
    this.resetCache();
  }
}
